app/src/main/java/com/mathewsachin/fategrandautomata/imaging/DroidCvPattern.kt

```diff
@@ -43,45 +65,77 @@ class DroidCvPattern(
 
     override fun close() {
         if (ownsMat) {
-            mat?.release()
+            mat.release()
         }
-
-        mat = null
+        alpha?.release()
     }
 
-    private fun resize(target: Mat, size: Size) {
+    private fun resize(source: Mat, target: Mat, size: Size) {
         Imgproc.resize(
-            mat, target,
+            source, target,
             CvSize(size.width.toDouble(), size.height.toDouble()),
             0.0, 0.0, Imgproc.INTER_AREA
         )
     }
 
     override fun resize(size: Size): IPattern {
-        val result = Mat()
-        resize(result, size)
-        return DroidCvPattern(result).tag(tag)
+        val resizedMat = Mat()
+        resize(mat, resizedMat, size)
+
+        val resizedAlpha = alpha?.let {
+            val target = Mat()
+            resize(it, target, size)
+            target
+        }
+
+        return DroidCvPattern(resizedMat, resizedAlpha).tag(tag)
+
     }
 
     override fun resize(target: IPattern, size: Size) {
         if (target is DroidCvPattern) {
-            resize(target.mat!!, size)
+            resize(mat, target.mat, size)
+            alpha?.let {
+                if (target.alpha == null) {
+                    target.alpha = Mat()
+                }
+                resize(it, target.alpha!!, size)
+            }
         }
 
         target.tag(tag)
     }
 
     private fun match(template: IPattern): Mat {
         val result = Mat()
-
         if (template is DroidCvPattern) {
             if (template.width <= width && template.height <= height) {
-                Imgproc.matchTemplate(
-                    mat,
-                    template.mat,
-                    result,
-                    Imgproc.TM_CCOEFF_NORMED
-                )
+                if (template.alpha != null) {
+                    Imgproc.matchTemplate(
+                        mat,
+                        template.mat,
+                        result,
+                        Imgproc.TM_CCOEFF_NORMED,
+                        template.alpha
+                    )
+                } else {
+                    Imgproc.matchTemplate(
+                        mat,
+                        template.mat,
+                        result,
+                        Imgproc.TM_CCOEFF_NORMED
+                    )
+                }
+
+                // in some cases, NaNs and Infinite values are part of the result
```

Comment:
I don't like this. We should check why we're getting Infinite and NaN in the first place.
These values didn't appear before so something could get bugged.

Reply:
This would also make some checks looking at the whole screen much slower 😢